src/main/java/org/openstreetmap/atlas/checks/base/ExternalDataFetcher.java

```diff
@@ -0,0 +1,174 @@
+                    }
+                    else
+                    {
+                        throw new CoreException("Unable to test existence of {}", uri, e);
+                    }
+                }
+            });
+            if (!exists)
+            {
+                if (!ExternalDataFetcher.this.silent)
+                {
+                    logger.warn("Fetcher: resource {} does not exist!", uri);
+                }
+                return Optional.empty();
+            }
+            return Optional.ofNullable(FileSystemHelper.resource(uri.toString(),
+                    ExternalDataFetcher.this.configuration));
+        }
+    }
+
+    private static final long serialVersionUID = 724339604023082195L;
+    private static final String GLOBAL_HADOOP_FILECACHE_NAMESPACE = "__HadoopExternalFileCache_global_namespace__";
+    private static final Logger logger = LoggerFactory.getLogger(ExternalDataFetcher.class);
+    /** Maximum number of retries for where an error occurs when getting a file */
+    private static final int RETRY_ATTEMPTS = 5;
+    /** The input folder path (same as the atlas file paths) */
+    private final String input;
+    /** The configuration used to create the cache */
+    private final Map<String, String> configuration;
+    /** The actual caching object */
+    private final transient ExternalDataResourceCache cache;
+    /**
+     * {@code true} implies that the caller does not want to log that files are not present
+     */
+    private boolean silent;
+
+    /**
+     * Create the fetcher to use for generic resources. The fetcher uses hadoop cache to reduce
+     * remote reads. See {@link ShardedIntegrityChecksSparkJob#atlasFetcher}.
+     *
+     * @param input
+     *            {@link String} input folder path
+     * @param configuration
+     *            {@link org.openstreetmap.atlas.generator.tools.spark.SparkJob} configuration map
+     */
+    public ExternalDataFetcher(final String input, final Map<String, String> configuration)
+    {
+        this.input = input;
+        this.configuration = configuration;
+        this.cache = new ExternalDataResourceCache();
+    }
+
+    @Override
+    public Optional<Resource> apply(final String string)
+    {
+        return this.getCache().get(this.getUri(string));
+    }
+
+    /**
+     * Make missed file messages silent (use when checking for files -- please log actual issues
+     * when this is used) Unfortunately, this does not suppress <i>all</i> messages from missed
+     * files.
+     *
+     * @param silent
+     *            {@code true} suppresses some logging messages from non-existent files
+     */
+    public void setSilent(final boolean silent)
+    {
+        this.silent = silent;
+    }
+
+    /**
+     * @return The resource cacher to use
+     */
+    @Nonnull
+    private ExternalDataResourceCache getCache()
+    {
+        return this.cache != null ? this.cache : new ExternalDataResourceCache();
```

Comment:
I think we need to assign this.cache if it's null then return it. Could be causing some cache misses

Reply:
I've made this change (I was originally looking at making `ExternalDataResourceCache` serializable, but that required changing `ConcurrentResourceCache` in an incompatible way).